src/lib/strategies/proxy-pki.js

```
@@ -125,112 +126,133 @@ function createUser(dn, acUser) {
-					updateUser(dn, { externalRoles: [], externalGroups: [] }).then(function(updatedUser) {
-						// Audit user signup
-						return auditService.audit('user updated from access checker', 'user', 'update', TeamMember.auditCopy(localUser), User.auditCopy(updatedUser)).then(() => updatedUser).then((result) => {
-							// Return the user
-							done(null, result);
-						}, done).done();
-					});
-				}
-				// Otherwise, just return the user
-				else {
-					done(null, localUser);
-				}
-			}
-			// If the user is known locally and in access checker, update their user info
-			else {
-				// Update the user only if we are not bypassing access checker
-				if (!localUser.bypassAccessCheck) {
-					updateUser(dn, copyACGroups(copyACRoles(copyACMetadata({}, acUser), acUser), acUser)).then(function(updatedUser) {
-						// Audit user signup
-						return auditService.audit('user updated from access checker', 'user', 'update', TeamMember.auditCopy(localUser), User.auditCopy(updatedUser)).then(function() {
-							return q(updatedUser);
-						});
-					}).then(function(result) {
-						// Return the user
-						done(null, result);
-					}, done).done();
-				}
-				else {
-					done(null, localUser);
+		try {
+			let proxiedUser = null;
+			let primaryUser = await handleUser(primaryUserDn, req);
+
+			if (primaryUser.canProxy && proxiedUserDn) {
+				proxiedUser = await handleUser(proxiedUserDn, req, true);
+
+				// Treat the proxied user account as if it's logging
+				// in by updating their lastLogin time.
+				if (!proxiedUser.lastLogin || proxiedUser.lastLogin + config.auth.sessionCookie.maxAge < Date.now()) {
+					proxiedUser = await User.findOneAndUpdate(
+						{ _id: proxiedUser._id },
+						{ lastLogin: Date.now() },
+						{ new: true, upsert: false }
+					);
 				}
 			}
 
-		}, function(err) {
-			// If there was an error, then something is broken and authentication has failed
-			return done(err);
-		});
-
+			if (proxiedUser === null) {
```

@jrassa thinking about some more use cases here... should we fail / throw an error if the request contains a proxied user DN, but the primary requestor is not allowed to proxy? That may be a more appropriate response for a server that is misconfigured, potentially avoiding any confusion about their ability to proxy.

---

On a side note, it looks like our unit tests for the node-rest-starter are ignoring errors from missing email templates in several places. It would be especially useful to have some test coverage around this piece of functionality (since it deals with authentication and authorization), so I'm going to work up a new branch to get those back online.